File: acquisition/build_index_immo.py
#!/usr/bin/env python3
"""INDEX ZERO-COPIE IMMO — geo province QC, 30 villes z∩m∩p.

Federation-first: l'index NE copie PAS la geometrie. Il reference feature_id (geoId)
+ no_lot pour qu'immo joigne aux PMTiles/cadastre de geo. Il ajoute code_zone
(point-in-polygon centroide sur grille zonage) + attrs batiment (jointure rôle foncier).

Inputs par ville (S3, lecture seule):
  1. normalized/qc-cadastre-lots/<slug>.geojson — features Polygon, props NO_LOT, geoId
  2. registry/role-foncier/<slug>.parquet — NO_LOT + role_* (jointure NO_LOT.replace(" ",""))
  3. normalized/ca-qc-zonage/<grid_slug>.geojson (18/30) — polygones props.zone_code

Sortie: registry/index-immo/<slug>.parquet (par ville) + registry/index-immo/manifest.json

ANTI-INVENTION ABSOLUE: code_zone=null si pas de grille OU centroide hors de tout
polygone; attrs batiment=null si pas de match rôle. Jamais deviner.

Idempotent: progress /tmp/index_progress.json (skip villes faites). Checkpoint par ville.
Foreground robuste, borné par --max-seconds (relancer pour continuer).
"""
import argparse
import io
import json
import os
import sys
import time
from collections import Counter

sys.path.insert(0, "/home/antoinefa/src/geo")

import boto3
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
from shapely.geometry import shape
from shapely.prepared import prep
from shapely.strtree import STRtree

S3ENV = "/home/antoinefa/src/_acquisition-shared/s3.env"
BUCKET = "sentropic-geo"
REGISTRY = "/home/antoinefa/src/_acquisition-shared/acquisition-registry.json"
GRIDS_MAP = "/home/antoinefa/src/_acquisition-shared/grids-slug-map.json"
PROG = "/tmp/index_progress.json"
SNAPSHOT = "2026-06-21"
SOURCE = "geo:cadastre-clip⋈role⋈zonage"

# Colonnes batiment issues du rôle foncier (jointure par no_lot normalisé).
ROLE_COLS = [
    "role_usage_cubf",
    "role_nb_etages_max",
    "role_annee_construction",
    "role_superficie_batiment_m2",
    "role_nb_logements",
    "role_valeur_immeuble",
]

# Schema de sortie (ordre + types pour le manifest).
OUT_SCHEMA = [
    ("feature_id", "string"),
    ("no_lot", "string"),
    ("code_zone", "string"),
    ("role_usage_cubf", "string"),
    ("role_nb_etages_max", "double"),
    ("role_annee_construction", "double"),
    ("role_superficie_batiment_m2", "double"),
    ("role_nb_logements", "double"),
    ("role_valeur_immeuble", "double"),
    ("_source", "string"),
    ("_snapshot", "string"),
]
# Types pyarrow explicites pour garder la stabilité du schéma malgré les colonnes
# entièrement nulles (anti-invention: une ville sans aucun match ne doit pas casser).
PA_FIELDS = [
    ("feature_id", pa.string()),
    ("no_lot", pa.string()),
    ("code_zone", pa.string()),
    ("role_usage_cubf", pa.string()),
    ("role_nb_etages_max", pa.float64()),
    ("role_annee_construction", pa.float64()),
    ("role_superficie_batiment_m2", pa.float64()),
    ("role_nb_logements", pa.float64()),
    ("role_valeur_immeuble", pa.float64()),
    ("_source", pa.string()),
    ("_snapshot", pa.string()),
]


def s3_client():
    env = {}
    with open(S3ENV) as f:
        for ln in f:
            ln = ln.strip()
            if "=" in ln and not ln.startswith("#"):
                k, v = ln.split("=", 1)
                env[k.strip()] = v.strip()
    return boto3.client(
        "s3",
        endpoint_url=env["S3_ENDPOINT"],
        region_name=env.get("S3_REGION", "fr-par"),
        aws_access_key_id=env["S3_ACCESS_KEY"],
        aws_secret_access_key=env["S3_SECRET_KEY"],
    )


def get_bytes(s3, key):
    return s3.get_object(Bucket=BUCKET, Key=key)["Body"].read()


def exists(s3, key):
    try:
        s3.head_object(Bucket=BUCKET, Key=key)
        return True
    except Exception:
        return False


def load_prog():
    try:
        return json.load(open(PROG))
    except Exception:
        return {"done": {}}


def save_prog(prog):
    json.dump(prog, open(PROG, "w"), indent=2, ensure_ascii=False)


def build_zone_index(s3, grid_slug):
    """Charge la grille zonage -> (STRtree des polygones, prepared, codes).

    Ne garde que les polygones avec un zone_code non-null (anti-invention).
    Retourne None si la grille est absente ou vide.
    """
    key = "normalized/ca-qc-zonage/%s.geojson" % grid_slug
    if not exists(s3, key):
        return None
    gj = json.loads(get_bytes(s3, key))
    geoms, codes = [], []
    for f in gj.get("features", []):
        zc = f.get("properties", {}).get("zone_code")
        g = f.get("geometry")
        if zc is None or zc == "" or g is None:
            continue
        try:
            sg = shape(g)
        except Exception:
            continue
        if sg.is_empty:
            continue
        geoms.append(sg)
        codes.append(str(zc))
    if not geoms:
        return None
    tree = STRtree(geoms)
    prepared = [prep(g) for g in geoms]
    return {"tree": tree, "prepared": prepared, "codes": codes, "n": len(geoms)}


def build_role_lookup(s3, slug):
    """Charge le parquet rôle -> dict {no_lot_normalisé: {role_col: val}}.

    Retourne ({}, 0) si le parquet est absent.
    """
    key = "registry/role-foncier/%s.parquet" % slug
    if not exists(s3, key):
        return {}, 0
    df = pq.read_table(io.BytesIO(get_bytes(s3, key))).to_pandas()
    cols = [c for c in ROLE_COLS if c in df.columns]
    df["_k"] = df["NO_LOT"].astype(str).str.replace(" ", "", regex=False)
    lookup = {}
    for rec in df[["_k"] + cols].to_dict("records"):
        k = rec.pop("_k")
        # Première occurrence gagne (clés uniques attendues).
        if k not in lookup:
            lookup[k] = rec
    return lookup, len(df)


def code_zone_for_point(zidx, pt):
    """Point-in-polygon: zone_code du 1er polygone contenant pt, sinon None."""
    if zidx is None:
        return None
    for idx in zidx["tree"].query(pt):
        if zidx["prepared"][idx].contains(pt):
            return zidx["codes"][idx]
    return None


def build_city(s3, slug, grid_slug):
    """Construit les lignes d'index pour une ville. Retourne (rows, stats)."""
    cad_key = "normalized/qc-cadastre-lots/%s.geojson" % slug
    if not exists(s3, cad_key):
        return None, {"error": "cadastre absent"}
    cad = json.loads(get_bytes(s3, cad_key))

    zidx = build_zone_index(s3, grid_slug) if grid_slug else None
    role_lookup, role_rows = build_role_lookup(s3, slug)

    rows = []
    with_zone = 0
    with_attrs = 0
    join_matched = 0
    for f in cad.get("features", []):
        p = f.get("properties", {})
        g = f.get("geometry")
        feature_id = p.get("geoId")
        no_lot = p.get("NO_LOT")
        # code_zone via centroide (representative_point garantit un point INTERNE).
        code_zone = None
        if g is not None:
            try:
                pt = shape(g).representative_point()
                code_zone = code_zone_for_point(zidx, pt)
            except Exception:
                code_zone = None
        if code_zone is not None:
            with_zone += 1
        # attrs batiment via jointure rôle (null si pas de match).
        k = str(no_lot).replace(" ", "") if no_lot is not None else None
        rec = role_lookup.get(k) if k is not None else None
        row = {
            "feature_id": feature_id,
            "no_lot": no_lot,
            "code_zone": code_zone,
            "_source": SOURCE,
            "_snapshot": SNAPSHOT,
        }
        if rec is not None:
            join_matched += 1
            any_attr = False
            for c in ROLE_COLS:
                v = rec.get(c)
                if v is not None and not (isinstance(v, float) and pd.isna(v)):
                    row[c] = v
                    any_attr = True
                else:
                    row[c] = None
            if any_attr:
                with_attrs += 1
        else:
            for c in ROLE_COLS:
                row[c] = None
        rows.append(row)

    n = len(rows)
    stats = {
        "lots": n,
        "with_code_zone": with_zone,
        "with_building_attrs": with_attrs,
        "join_matched": join_matched,
        "code_zone_pct": round(100 * with_zone / n, 2) if n else 0.0,
        "building_pct": round(100 * with_attrs / n, 2) if n else 0.0,
        "join_pct": round(100 * join_matched / n, 2) if n else 0.0,
        "has_grid": bool(zidx),
        "grid_slug": grid_slug,
        "zone_polys": zidx["n"] if zidx else 0,
        "role_rows": role_rows,
    }
    return rows, stats


def write_parquet(rows, path):
    # Normalise les types role_usage_cubf en string (peut être int dans le rôle).
    arrays = {}
    for name, _ in PA_FIELDS:
        arrays[name] = [r.get(name) for r in rows]
    # role_usage_cubf -> string
    arrays["role_usage_cubf"] = [
        None if v is None or (isinstance(v, float) and pd.isna(v)) else str(int(v)) if isinstance(v, float) and v.is_integer() else str(v)
        for v in arrays["role_usage_cubf"]
    ]
    schema = pa.schema(PA_FIELDS)
    cols = {}
    for name, typ in PA_FIELDS:
        cols[name] = pa.array(arrays[name], type=typ)
    table = pa.table(cols, schema=schema)
    pq.write_table(table, path, compression="zstd")
    return table.num_rows


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--max-seconds", type=int, default=540)
    ap.add_argument("--only", help="slug unique (test)")
    ap.add_argument("--no-upload", action="store_true", help="dry-run local, pas d'upload S3")
    ap.add_argument("--force", action="store_true", help="ignore le progress (rebuild)")
    a = ap.parse_args()

    s3 = s3_client()
    registry = json.load(open(REGISTRY))
    grids = json.load(open(GRIDS_MAP))
    cities = list(registry.keys())
    if a.only:
        cities = [c for c in cities if c == a.only]

    prog = load_prog()
    if a.force:
        prog = {"done": {}}
    os.makedirs("/tmp/index_immo", exist_ok=True)
    t0 = time.time()

    for slug in cities:
        if slug in prog["done"] and not a.force:
            print("SKIP(done) %s" % slug)
            continue
        if time.time() - t0 > a.max_seconds:
            print("STOP wall-clock; relancer pour continuer")
            break
        grid_slug = grids.get(slug)
        try:
            rows, stats = build_city(s3, slug, grid_slug)
        except Exception as e:
            print("FAIL-BUILD %-44s %s" % (slug, e))
            time.sleep(1)
            continue
        if rows is None:
            print("FAIL %-44s %s" % (slug, stats.get("error")))
            continue
        path = "/tmp/index_immo/%s.parquet" % slug
        nrows = write_parquet(rows, path)
        if not a.no_upload:
            s3.upload_file(path, BUCKET, "registry/index-immo/%s.parquet" % slug)
        print(
            "OK  %-44s lots=%-6d code_zone=%5.1f%% building=%5.1f%% join=%5.1f%% grid=%s"
            % (slug, nrows, stats["code_zone_pct"], stats["building_pct"], stats["join_pct"], "Y" if stats["has_grid"] else "-")
        )
        prog["done"][slug] = stats
        save_prog(prog)
        time.sleep(0.2)

    print("=== chunk fin (%d/%d done) ===" % (len(prog["done"]), len(registry)))


if __name__ == "__main__":
    main()

File: acquisition/src/build-index-immo.ts
/**
 * build-index-immo.ts — INDEX ZERO-COPIE IMMO (geo province QC).
 *
 * Port fidèle de `acquisition/build_index_immo.py`. Federation-first : l'index
 * NE copie PAS la géométrie ; il référence feature_id (geoId) + no_lot, ajoute
 * code_zone (point-in-polygon centroïde sur grille zonage) + attrs bâtiment
 * (jointure rôle foncier par no_lot normalisé).
 *
 * Ce module exporte les fonctions de join réutilisées par
 * `cadastre-index-province.ts` (buildCity / writeIndexParquet / buildZoneIndex),
 * comme le .py réutilise `build_index_immo as core`.
 *
 * ANTI-INVENTION : code_zone=null si pas de grille OU centroïde hors polygone ;
 * attrs bâtiment=null si pas de match rôle. Jamais deviner.
 */
import type { S3Client } from "@aws-sdk/client-s3";
import type { Feature, Geometry, Polygon, MultiPolygon, Position } from "geojson";

import { BUCKET as DEFAULT_BUCKET, exists, getBytes, getJson } from "./lib/s3.js";
import { representativePoint, strictPointInPolygon } from "./lib/geo.js";
import { readParquetRowsFromBuffer } from "./lib/parquet-read.js";
import { writeIndexImmoParquet, INDEX_IMMO_COLUMNS, type IndexImmoRow } from "./lib/parquet.js";

export const BUCKET = DEFAULT_BUCKET;
export const REGISTRY = "/home/antoinefa/src/_acquisition-shared/acquisition-registry.json";
export const GRIDS_MAP = "/home/antoinefa/src/_acquisition-shared/grids-slug-map.json";
export const SNAPSHOT = "2026-06-21";
export const SOURCE = "geo:cadastre-clip⋈role⋈zonage";

/** Colonnes bâtiment issues du rôle (jointure par no_lot normalisé). */
export const ROLE_COLS = [
  "role_usage_cubf",
  "role_nb_etages_max",
  "role_annee_construction",
  "role_superficie_batiment_m2",
  "role_nb_logements",
  "role_valeur_immeuble",
] as const;

/** Schéma de sortie (ordre + types pour le manifest), miroir de OUT_SCHEMA. */
export const OUT_SCHEMA: [string, string][] = [
  ["feature_id", "string"],
  ["no_lot", "string"],
  ["code_zone", "string"],
  ["role_usage_cubf", "string"],
  ["role_nb_etages_max", "double"],
  ["role_annee_construction", "double"],
  ["role_superficie_batiment_m2", "double"],
  ["role_nb_logements", "double"],
  ["role_valeur_immeuble", "double"],
  ["_source", "string"],
  ["_snapshot", "string"],
];

type GeoFeature = Feature<Geometry, Record<string, unknown> | null>;
type FeatureCollection = { features?: GeoFeature[] };

export interface ZoneIndex {
  polys: (Polygon | MultiPolygon)[];
  codes: string[];
  n: number;
}

/**
 * Charge la grille zonage -> { polys, codes }. Ne garde que les polygones avec
 * un zone_code non-null (anti-invention). Retourne null si absente/vide.
 */
export async function buildZoneIndex(
  s3: S3Client,
  gridSlug: string,
): Promise<ZoneIndex | null> {
  const key = `normalized/ca-qc-zonage/${gridSlug}.geojson`;
  if (!(await exists(s3, key))) return null;
  const gj = (await getJson(s3, key)) as FeatureCollection;
  const polys: (Polygon | MultiPolygon)[] = [];
  const codes: string[] = [];
  for (const f of gj.features ?? []) {
    const zc = (f.properties ?? {})["zone_code"];
    const g = f.geometry;
    if (zc === null || zc === undefined || zc === "" || !g) continue;
    if (g.type !== "Polygon" && g.type !== "MultiPolygon") continue;
    polys.push(g as Polygon | MultiPolygon);
    codes.push(String(zc));
  }
  if (polys.length === 0) return null;
  return { polys, codes, n: polys.length };
}

/**
 * Charge le parquet rôle -> { lookup: {no_lot_normalisé: {role_col: val}}, n }.
 * Retourne lookup vide si absent / dégénéré (pas de NO_LOT).
 */
export async function buildRoleLookup(
  s3: S3Client,
  slug: string,
): Promise<{ lookup: Map<string, Record<string, unknown>>; roleRows: number }> {
  const key = `registry/role-foncier/${slug}.parquet`;
  if (!(await exists(s3, key))) return { lookup: new Map(), roleRows: 0 };
  const buf = await getBytes(s3, key);
  const rows = await readParquetRowsFromBuffer(buf);
  if (rows.length === 0 || !("NO_LOT" in (rows[0] ?? {}))) {
    return { lookup: new Map(), roleRows: 0 };
  }
  const cols = ROLE_COLS.filter((c) => c in (rows[0] ?? {}));
  const lookup = new Map<string, Record<string, unknown>>();
  for (const r of rows) {
    const k = String(r["NO_LOT"] ?? "").replace(/ /g, "");
    if (lookup.has(k)) continue; // première occurrence gagne
    const rec: Record<string, unknown> = {};
    for (const c of cols) rec[c] = r[c];
    lookup.set(k, rec);
  }
  return { lookup, roleRows: rows.length };
}

/** code_zone du 1er polygone contenant pt (strict), sinon null. */
export function codeZoneForPoint(zidx: ZoneIndex | null, pt: Position): string | null {
  if (!zidx) return null;
  for (let i = 0; i < zidx.polys.length; i++) {
    if (strictPointInPolygon(pt, zidx.polys[i]!)) return zidx.codes[i]!;
  }
  return null;
}

export interface CityStats {
  lots: number;
  with_code_zone: number;
  with_building_attrs: number;
  join_matched: number;
  code_zone_pct: number;
  building_pct: number;
  join_pct: number;
  has_grid: boolean;
  grid_slug: string | null;
  zone_polys: number;
  role_rows: number;
  error?: string;
}

const isNa = (v: unknown): boolean =>
  v === null || v === undefined || (typeof v === "number" && Number.isNaN(v));

/** Construit les lignes d'index pour une ville. */
export async function buildCity(
  s3: S3Client,
  slug: string,
  gridSlug: string | null,
): Promise<{ rows: IndexImmoRow[] | null; stats: Partial<CityStats> }> {
  const cadKey = `normalized/qc-cadastre-lots/${slug}.geojson`;
  if (!(await exists(s3, cadKey))) return { rows: null, stats: { error: "cadastre absent" } };
  const cad = (await getJson(s3, cadKey)) as FeatureCollection;

  const zidx = gridSlug ? await buildZoneIndex(s3, gridSlug) : null;
  const { lookup, roleRows } = await buildRoleLookup(s3, slug);

  const rows: IndexImmoRow[] = [];
  let withZone = 0;
  let withAttrs = 0;
  let joinMatched = 0;
  for (const f of cad.features ?? []) {
    const p = f.properties ?? {};
    const g = f.geometry;
    const featureId = p["geoId"] ?? null;
    const noLot = p["NO_LOT"] ?? null;

    let codeZone: string | null = null;
    if (g) {
      const pt = representativePoint(g);
      if (pt) codeZone = codeZoneForPoint(zidx, pt);
    }
    if (codeZone !== null) withZone++;

    const k = noLot !== null && noLot !== undefined ? String(noLot).replace(/ /g, "") : null;
    const rec = k !== null ? lookup.get(k) : undefined;

    const row: Record<string, unknown> = {
      feature_id: featureId,
      no_lot: noLot,
      code_zone: codeZone,
      _source: SOURCE,
      _snapshot: SNAPSHOT,
    };
    if (rec) {
      joinMatched++;
      let anyAttr = false;
      for (const c of ROLE_COLS) {
        const v = rec[c];
        if (!isNa(v)) {
          row[c] = v;
          anyAttr = true;
        } else row[c] = null;
      }
      if (anyAttr) withAttrs++;
    } else {
      for (const c of ROLE_COLS) row[c] = null;
    }
    rows.push(row as IndexImmoRow);
  }

  const n = rows.length;
  const stats: CityStats = {
    lots: n,
    with_code_zone: withZone,
    with_building_attrs: withAttrs,
    join_matched: joinMatched,
    code_zone_pct: n ? round2((100 * withZone) / n) : 0,
    building_pct: n ? round2((100 * withAttrs) / n) : 0,
    join_pct: n ? round2((100 * joinMatched) / n) : 0,
    has_grid: !!zidx,
    grid_slug: gridSlug,
    zone_polys: zidx ? zidx.n : 0,
    role_rows: roleRows,
  };
  return { rows, stats };
}

/**
 * Écrit les lignes en parquet (schéma index-immo fixe). role_usage_cubf est
 * normalisé en string (int -> "1000"), comme le .py.
 */
export async function writeParquet(rows: IndexImmoRow[], path: string): Promise<number> {
  const out: IndexImmoRow[] = rows.map((r) => {
    const o: Record<string, unknown> = {};
    for (const c of INDEX_IMMO_COLUMNS) o[c] = r[c];
    const cubf = o["role_usage_cubf"];
    if (isNa(cubf)) o["role_usage_cubf"] = null;
    else if (typeof cubf === "number")
      o["role_usage_cubf"] = Number.isInteger(cubf) ? String(cubf) : String(cubf);
    else o["role_usage_cubf"] = String(cubf);
    return o as IndexImmoRow;
  });
  return writeIndexImmoParquet(out, path);
}

function round2(x: number): number {
  return Math.round(x * 100) / 100;
}
